pkg/rc/consul_backed.go

```diff
@@ -0,0 +1,133 @@
+package rc
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/square/p2/Godeps/_workspace/src/github.com/hashicorp/consul/api"
+
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/pods"
+)
+
+type consulKV interface {
+	Put(*api.KVPair, *api.WriteOptions) (*api.WriteMeta, error)
+}
+
+type consulRc struct {
+	innerRc ReplicationController
+	kv      consulKV
+	kpStore kpStore
+}
+
+type kvPair struct {
+	key   string
+	value []byte
+}
+
+// consulRc implements the ReplicationController interface.
+var _ ReplicationController = &consulRc{}
+
+func NewConsulBacked(consulClient *api.Client, innerRc ReplicationController) (*consulRc, error) {
+	rc := &consulRc{
+		innerRc: innerRc,
+		kv:      consulClient.KV(),
+		kpStore: kp.NewFromClient(consulClient),
+	}
+
+	// Store everything related to the RC into consul, bail if errored.
+	_, err := rc.kpStore.SetPod(rc.consulPrefix()+"/pod_manifest", *innerRc.Manifest())
```

Comment:
Or just enumerated in this pkg since this owns RCs. One thing I like about the `pods` package is that there are no references to Consul inside it - `kp` maps Consul values onto `pod` objects